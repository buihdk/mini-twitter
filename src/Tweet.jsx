import React from 'react';
import './App.css';

export default class Tweet extends React.Component {
  render() {
    let tweet = this.props.tweet;
    const date = new Date().toLocaleString();
    return (
      <div className="tweet media my-3">
        <img className="avatar-tweet mr-3" src="https://pbs.twimg.com/profile_images/553467511211970560/nBE77dF0_400x400.jpeg" alt="avatar-tweet" width="200px"/>
        <div className="media-body">
          <h5 className="mt-0">Khoa Bui <span class="sub-text">@buihdk {date}</span></h5>
          <p>{tweet.text}</p>
          <i class="fas fa-retweet" onClick={() => this.props.handleRetweet(tweet)}></i>&nbsp;&nbsp;&nbsp;&nbsp;
          <i className={tweet.liked ? "fas fa-heart liked" : "fas fa-heart"} onClick={() => this.props.handleLike(tweet)}/>&nbsp;&nbsp;&nbsp;&nbsp;
          <i className="far fa-trash-alt" onClick={() => this.props.handleDelete(tweet)}/>
        </div>
      </div>
    )
  }
}