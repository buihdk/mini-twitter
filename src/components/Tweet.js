import React from 'react';
import '../styles/App.css';

const Tweet = () => {
  let tweet = this.props.tweet;
  return (
    <div className='tweet media my-3'>
      <img className='avatar-tweet mr-3' src="https://pbs.twimg.com/profile_images/553467511211970560/nBE77dF0_400x400.jpeg" alt='avatar-tweet' width='200px'/>
      <div className='media-body'>
        <h5 className='mt-0'>Khoa Bui <span className='sub-text'>@buihdk {tweet.date}</span></h5>
        <p>{tweet.text}</p>
        <i className='fas fa-retweet' onClick={() => this.props.handleRetweet(tweet)}></i>&nbsp;&nbsp;&nbsp;&nbsp;
        <i className={tweet.liked ? 'fas fa-heart liked' : 'fas fa-heart'} onClick={() => this.props.handleLike(tweet)}/>&nbsp;&nbsp;&nbsp;&nbsp;
        <i className='far fa-trash-alt' onClick={() => this.props.handleDelete(tweet)}/>
      </div>
    </div>
  );
};

export default Tweet;