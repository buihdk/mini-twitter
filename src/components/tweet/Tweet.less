.tweet {
  background-color: white;
  padding: 1rem;
  box-shadow: 0 2px 3px rgba(10, 10, 10, 0.1), 0 0 0 1px rgba(10, 10, 10, 0.1);
  word-wrap: break-word;
}

.avatar-tweet {
  width: 64px;
  border-radius: 50%;
}

.fa-retweet,
.fa-heart,
.fa-trash-alt,
.sub-text {
  color: lightgray;
  font-size: 16px;
}

.fa-retweet:hover,
.fa-heart:hover,
.fa-trash-alt:hover {
  color: gray;
  cursor: pointer;
}

.fa-heart.liked {
  color: red;
}