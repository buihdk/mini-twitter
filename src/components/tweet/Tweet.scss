$box-shadow-color: rgba(10, 10, 10, 0.1);

.tweet {
  background-color: white;
  padding: 1rem;
  box-shadow: 0 2px 3px $box-shadow-color, 0 0 0 1px $box-shadow-color;
  word-wrap: break-word;
}

.avatar-tweet {
  width: 64px;
  border-radius: 50%;
}

.sub-text  {
  color: lightgray;
  font-size: 16px;
}

.media-body > svg {
  color: lightgray;
  font-size: 16px;
  margin-right: 8px;
  &:hover {
    color: gray;
    cursor: pointer;
  }
}

.liked {
  color: red !important;
}