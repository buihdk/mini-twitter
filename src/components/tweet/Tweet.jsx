import React, { memo } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { Media, Image } from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faRetweet,
  faHeart,
  faTrashAlt,
} from '@fortawesome/free-solid-svg-icons';

import './Tweet.scss';

const Tweet = ({ tweet, handleRetweet, handleLike, handleDelete }) => (
  <Media className="tweet my-3">
    <Image
      roundedCircle
      width="64px"
      className="mr-3"
      src="https://pbs.twimg.com/profile_images/553467511211970560/nBE77dF0_400x400.jpeg"
    />
    <Media.Body>
      <h5 className="mt-0">
        {`Khoa Bui `}
        <span className="sub-text">
          {`@buihdk ${moment(tweet.date).fromNow()}`}
        </span>
      </h5>
      <p>{tweet.text}</p>
      <FontAwesomeIcon icon={faRetweet} onClick={() => handleRetweet(tweet)} />
      <FontAwesomeIcon
        icon={faHeart}
        onClick={() => handleLike(tweet)}
        className={tweet.liked ? 'liked' : ''}
      />
      <FontAwesomeIcon icon={faTrashAlt} onClick={() => handleDelete(tweet)} />
    </Media.Body>
  </Media>
);

Tweet.propTypes = {
  tweet: PropTypes.shape({
    text: PropTypes.string,
    liked: PropTypes.bool,
    date: PropTypes.string,
  }),
  handleRetweet: PropTypes.func,
  handleLike: PropTypes.func,
  handleDelete: PropTypes.func,
};

Tweet.defaultProps = {
  tweet: {
    text: '',
    liked: false,
    date: '',
  },
  handleRetweet: () => {},
  handleLike: () => {},
  handleDelete: () => {},
};

export default memo(Tweet);
