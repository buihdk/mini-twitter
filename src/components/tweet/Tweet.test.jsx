import React from 'react';
import { shallow, mount } from 'enzyme';
import toJSON from 'enzyme-to-json';

import Tweet from './Tweet';

const tweet = {
  text: 'test',
  liked: false,
  date: new Date().toISOString(),
};
const likedTweet = { ...tweet, liked: true };
const handleRetweet = jest.fn();
const handleLike = jest.fn();
const handleDelete = jest.fn();

let wrapper;

afterEach(() => {
  wrapper = null;
});

afterAll(() => {
  handleRetweet.mockClear();
  handleLike.mockClear();
  handleDelete.mockClear();
});

describe('Tweet', () => {
  describe('empty props', () => {
    test('renders without crashing', () => {
      wrapper = shallow(<Tweet />);

      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('with props', () => {
    test('renders without crashing', () => {
      wrapper = shallow(<Tweet tweet={tweet} />);

      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders liked tweet without crashing', () => {
      wrapper = shallow(<Tweet tweet={likedTweet} />);

      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('handleRetweet', () => {
      wrapper = mount(<Tweet tweet={tweet} handleRetweet={handleRetweet} />);
      wrapper.find('.fa-retweet').simulate('click');

      expect(handleRetweet).toBeCalledWith(tweet);
    });

    test('handleLike', () => {
      wrapper = mount(<Tweet tweet={tweet} handleLike={handleLike} />);
      wrapper.find('.fa-heart').simulate('click');

      expect(handleLike).toBeCalledWith(tweet);
    });

    test('handleLike', () => {
      wrapper = mount(<Tweet tweet={tweet} handleDelete={handleDelete} />);
      wrapper.find('.fa-trash-alt').simulate('click');

      expect(handleDelete).toBeCalledWith(tweet);
    });
  });
});
