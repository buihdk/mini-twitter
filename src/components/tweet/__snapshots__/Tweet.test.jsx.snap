// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tweet empty props renders without crashing 1`] = `
<Media
  className="tweet my-3"
>
  <Bootstrap(Image)
    className="mr-3"
    roundedCircle={true}
    src="https://pbs.twimg.com/profile_images/553467511211970560/nBE77dF0_400x400.jpeg"
    width="64px"
  />
  <MediaBody>
    <h5
      className="mt-0"
    >
      Khoa Bui 
      <span
        className="sub-text"
      >
        @buihdk Invalid date
      </span>
    </h5>
    <p />
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            640,
            512,
            Array [],
            "f079",
            "M629.657 343.598L528.971 444.284c-9.373 9.372-24.568 9.372-33.941 0L394.343 343.598c-9.373-9.373-9.373-24.569 0-33.941l10.823-10.823c9.562-9.562 25.133-9.34 34.419.492L480 342.118V160H292.451a24.005 24.005 0 0 1-16.971-7.029l-16-16C244.361 121.851 255.069 96 276.451 96H520c13.255 0 24 10.745 24 24v222.118l40.416-42.792c9.285-9.831 24.856-10.054 34.419-.492l10.823 10.823c9.372 9.372 9.372 24.569-.001 33.941zm-265.138 15.431A23.999 23.999 0 0 0 347.548 352H160V169.881l40.416 42.792c9.286 9.831 24.856 10.054 34.419.491l10.822-10.822c9.373-9.373 9.373-24.569 0-33.941L144.971 67.716c-9.373-9.373-24.569-9.373-33.941 0L10.343 168.402c-9.373 9.373-9.373 24.569 0 33.941l10.822 10.822c9.562 9.562 25.133 9.34 34.419-.491L96 169.881V392c0 13.255 10.745 24 24 24h243.549c21.382 0 32.09-25.851 16.971-40.971l-16.001-16z",
          ],
          "iconName": "retweet",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f004",
            "M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z",
          ],
          "iconName": "heart",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f2ed",
            "M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z",
          ],
          "iconName": "trash-alt",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </MediaBody>
</Media>
`;

exports[`Tweet with props renders liked tweet without crashing 1`] = `
<Media
  className="tweet my-3"
>
  <Bootstrap(Image)
    className="mr-3"
    roundedCircle={true}
    src="https://pbs.twimg.com/profile_images/553467511211970560/nBE77dF0_400x400.jpeg"
    width="64px"
  />
  <MediaBody>
    <h5
      className="mt-0"
    >
      Khoa Bui 
      <span
        className="sub-text"
      >
        @buihdk a few seconds ago
      </span>
    </h5>
    <p>
      test
    </p>
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            640,
            512,
            Array [],
            "f079",
            "M629.657 343.598L528.971 444.284c-9.373 9.372-24.568 9.372-33.941 0L394.343 343.598c-9.373-9.373-9.373-24.569 0-33.941l10.823-10.823c9.562-9.562 25.133-9.34 34.419.492L480 342.118V160H292.451a24.005 24.005 0 0 1-16.971-7.029l-16-16C244.361 121.851 255.069 96 276.451 96H520c13.255 0 24 10.745 24 24v222.118l40.416-42.792c9.285-9.831 24.856-10.054 34.419-.492l10.823 10.823c9.372 9.372 9.372 24.569-.001 33.941zm-265.138 15.431A23.999 23.999 0 0 0 347.548 352H160V169.881l40.416 42.792c9.286 9.831 24.856 10.054 34.419.491l10.822-10.822c9.373-9.373 9.373-24.569 0-33.941L144.971 67.716c-9.373-9.373-24.569-9.373-33.941 0L10.343 168.402c-9.373 9.373-9.373 24.569 0 33.941l10.822 10.822c9.562 9.562 25.133 9.34 34.419-.491L96 169.881V392c0 13.255 10.745 24 24 24h243.549c21.382 0 32.09-25.851 16.971-40.971l-16.001-16z",
          ],
          "iconName": "retweet",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <FontAwesomeIcon
      border={false}
      className="liked"
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f004",
            "M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z",
          ],
          "iconName": "heart",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f2ed",
            "M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z",
          ],
          "iconName": "trash-alt",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </MediaBody>
</Media>
`;

exports[`Tweet with props renders without crashing 1`] = `
<Media
  className="tweet my-3"
>
  <Bootstrap(Image)
    className="mr-3"
    roundedCircle={true}
    src="https://pbs.twimg.com/profile_images/553467511211970560/nBE77dF0_400x400.jpeg"
    width="64px"
  />
  <MediaBody>
    <h5
      className="mt-0"
    >
      Khoa Bui 
      <span
        className="sub-text"
      >
        @buihdk a few seconds ago
      </span>
    </h5>
    <p>
      test
    </p>
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            640,
            512,
            Array [],
            "f079",
            "M629.657 343.598L528.971 444.284c-9.373 9.372-24.568 9.372-33.941 0L394.343 343.598c-9.373-9.373-9.373-24.569 0-33.941l10.823-10.823c9.562-9.562 25.133-9.34 34.419.492L480 342.118V160H292.451a24.005 24.005 0 0 1-16.971-7.029l-16-16C244.361 121.851 255.069 96 276.451 96H520c13.255 0 24 10.745 24 24v222.118l40.416-42.792c9.285-9.831 24.856-10.054 34.419-.492l10.823 10.823c9.372 9.372 9.372 24.569-.001 33.941zm-265.138 15.431A23.999 23.999 0 0 0 347.548 352H160V169.881l40.416 42.792c9.286 9.831 24.856 10.054 34.419.491l10.822-10.822c9.373-9.373 9.373-24.569 0-33.941L144.971 67.716c-9.373-9.373-24.569-9.373-33.941 0L10.343 168.402c-9.373 9.373-9.373 24.569 0 33.941l10.822 10.822c9.562 9.562 25.133 9.34 34.419-.491L96 169.881V392c0 13.255 10.745 24 24 24h243.549c21.382 0 32.09-25.851 16.971-40.971l-16.001-16z",
          ],
          "iconName": "retweet",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            512,
            512,
            Array [],
            "f004",
            "M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z",
          ],
          "iconName": "heart",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
    <FontAwesomeIcon
      border={false}
      className=""
      fixedWidth={false}
      flip={null}
      icon={
        Object {
          "icon": Array [
            448,
            512,
            Array [],
            "f2ed",
            "M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z",
          ],
          "iconName": "trash-alt",
          "prefix": "fas",
        }
      }
      inverse={false}
      listItem={false}
      mask={null}
      onClick={[Function]}
      pull={null}
      pulse={false}
      rotation={null}
      size={null}
      spin={false}
      swapOpacity={false}
      symbol={false}
      title=""
      transform={null}
    />
  </MediaBody>
</Media>
`;
